### ISSUE TITLE
iptables FORWARD chain DROP policy breaks workers

### ISSUE THREAD

**Issue by jpf91** [concourse/concourse#897]
# Bug Report

The default `w--forward` rule seems to be insufficient when using a `FORWARD` chain which defaults to the `DROP` policy. A `FORWARD` chain with a default `ACCEPT` policy is a security risk, so I'd prefer to be able to use the `DROP` policy.

Worker and Web are installed on the same machine and started like this:
```
/usr/local/bin/concourse web --basic-auth-username "user" --basic-auth-password 'pass' --session-signing-key /etc/concourse/session_signing_key --tsa-host-key /etc/concourse/host_key --tsa-authorized-keys /etc/concourse/authorized_worker_keys --external-url "http://EXTIP:8080" --postgres-data-source 'postgres://concourse:concourse@127.0.0.1:5432/atc?sslmode=disable'

/usr/local/bin/concourse worker --work-dir /var/lib/concourse --tsa-host localhost --tsa-public-key /etc/concourse/host_key.pub --tsa-worker-private-key /etc/concourse/worker_key_builder_vn1
```

iptables rules:
```
*filter
:INPUT DROP [0:0]
:FORWARD DROP [0:0]
:OUTPUT ACCEPT [369:40109]
:TCP - [0:0]
:UDP - [0:0]
-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A INPUT -m conntrack --ctstate INVALID -j DROP
-A INPUT -p icmp -m icmp --icmp-type 8 -m conntrack --ctstate NEW -j ACCEPT
-A INPUT -i ens18 -p udp -m conntrack --ctstate NEW -j UDP
-A INPUT -i ens18 -p tcp -m tcp --tcp-flags FIN,SYN,RST,ACK SYN -m conntrack --ctstate NEW -j TCP
-A INPUT -i ens18 -p udp -j REJECT --reject-with icmp-port-unreachable
-A INPUT -i ens18 -p tcp -j REJECT --reject-with tcp-reset
-A INPUT -i ens18 -j REJECT --reject-with icmp-proto-unreachable
-A TCP -p tcp -m tcp --dport 8080 -j ACCEPT

COMMIT
# Completed on Mon Jan  9 14:46:29 2017
```
Some debugging (slightly modified rules compared to above for debugging) shows this:
```      
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  *      w+      EXTIP       0.0.0.0/0            ctstate NEW
    3   180 w--forward  all  --  w+     *       0.0.0.0/0            0.0.0.0/0           
    3   180 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain w--default (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED

Chain w--forward (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     all  --  ens18  *       0.0.0.0/0            0.0.0.0/0           
    3   180 w--instance-7cvnoioiv4f  all  --  wbrdg-0afe0000 *       10.254.0.2           0.0.0.0/0           [goto] 
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain w--instance-7cvnoioiv4f (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     all  --  *      *       10.254.0.0/30        10.254.0.0/30       
    3   180 w--default  all  --  *      *       0.0.0.0/0            0.0.0.0/0           [goto] 
```
So it looks as if some packets are process by `w--forward ==> w--instance-7cvnoioiv4f => w--default` but not accepted. This then causes a job to be stuck in the pending state forever.

A workaround is to use these rules:
```
-A FORWARD -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -s EXTIP/32 -o w+ -m conntrack --ctstate NEW -j ACCEPT
-A FORWARD -i w+ -m conntrack --ctstate NEW -j ACCEPT
```
as the `w--forward ` is added at the end of the chain, the last rule above will 'steal' all packages from that rule but as long as `w--forward` only ACCEPTs packets this is not a problem.

BTW: Some documentation about firewall setup / how concourse uses bridges/iptables would be very appreciated. Many iptables configurations are set up for all interfaces by default and will therefore also apply to the bridge interfaces used by concourse. This is something that needs to be considered by sysadmins ;-)

* concourse 2.5.1, binary release, on ubuntu 16.04.1 LTS
* Simple stateful firewall, according to https://wiki.archlinux.org/index.php/simple_stateful_firewall
* Not sure if this ever worked, I'm new to concourse CI ;-)

**Comment by pn-santos** [comment]
I've come across this too with an extra issue, the machines I install workers in come with a default DROP policy for the FORWARD and INPUT chain and the last rule of the chain is a REJECT. 

When starting the worker concourse adds the  `w--forward` at the end of the FORWARD chain, which means that packets will be dropped before they reach the `w--forward` rule. However even if the `w--forward` rule is moved up, any NEW connections will be dropped (since the `w--default` rule doesn't pick up packets with cstate NEW)

Initial INPUT and FORWARD chains:
```
Chain INPUT (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    (...)
    0     0 LOG        all  --  *      *       0.0.0.0/0            0.0.0.0/0            limit: avg 5/min burst 5 LOG flags 0 level 7 prefix "iptables INPUT denied: "
    0     0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-port-unreachable

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 LOG        all  --  *      *       0.0.0.0/0            0.0.0.0/0            limit: avg 5/min burst 5 LOG flags 0 level 7 prefix "iptables FORWARD denied: "
    0     0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-port-unreachable
```

After the concourse worker is started:
```
Chain INPUT (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    (...)
    0     0 LOG        all  --  *      *       0.0.0.0/0            0.0.0.0/0            limit: avg 5/min burst 5 LOG flags 0 level 7 prefix "iptables INPUT denied: "
    0     0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-port-unreachable
    0     0 w--input   all  --  w+     *       0.0.0.0/0            0.0.0.0/0

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 LOG        all  --  *      *       0.0.0.0/0            0.0.0.0/0            limit: avg 5/min burst 5 LOG flags 0 level 7 prefix "iptables FORWARD denied: "
    0     0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-port-unreachable
    0     0 w--forward  all  --  w+     *       0.0.0.0/0            0.0.0.0/0

Chain w--default (0 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED

Chain w--forward (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     all  --  eth1   *       0.0.0.0/0            0.0.0.0/0
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain w--input (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     all  --  eth1   *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0
```
After moving the `w--forward` and `w--input` rule up and trying to run a job it's easy to see that requests are being blocked
```
Apr 21 10:01:14 n0 kernel: iptables FORWARD denied: IN=wbrdg-c0a80000 OUT=eth1 MAC=66:b8:b0:60:92:a4:82:52:67:81:a5:88:08:00 SRC=192.168.0.2 DST=10.34.104.11 LEN=66 TOS=0x00 PREC=0x00 TTL=63 ID=58362 DF PROTO=UDP SPT=38584 DPT=53 LEN=46
```
So my first question is shouldn't the `w--default` rule be `ctstate NEW,RELATED,ESTABLISHED` ?
This will not completely solve the problem though since inbound forwarding is also blocked (after adding `cstate NEW` to the `w--default` rule)
```
Apr 21 10:08:48 n0 kernel: iptables FORWARD denied: IN=eth1 OUT=wbrdg-c0a80004 MAC=00:0c:29:42:14:01:00:10:db:ff:10:00:08:00 SRC=10.34.104.11 DST=192.168.0.6 LEN=150 TOS=0x00 PREC=0x00 TTL=62 ID=39046 PROTO=UDP SPT=53 DPT=38440 LEN=130
```

As far as I can tell the following config solves the forwarding problem but I'm not network expert so not really sure if it makes sense or is the right solution.
```
Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
  884 66730 w--forward  all  --  w+     *       0.0.0.0/0            0.0.0.0/0
 1512 2063K w--forward  all  --  *      w+      0.0.0.0/0            0.0.0.0/0
    0     0 LOG        all  --  *      *       0.0.0.0/0            0.0.0.0/0            limit: avg 5/min burst 5 LOG flags 0 level 7 prefix "iptables FORWARD denied: "
    0     0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-port-unreachable

Chain w--default (7 references)
 pkts bytes target     prot opt in     out     source               destination
  884 66730 ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate NEW,RELATED,ESTABLISHED

Chain w--forward (2 references)
 pkts bytes target     prot opt in     out     source               destination
 1512 2063K ACCEPT     all  --  eth1   *       0.0.0.0/0            0.0.0.0/0
    (... forwarding rules for each container instance ...)
```

So in summary:

1. With a default DROP FORWARD policy, concourse generated rules are not sufficient to enable traffic forwarding between host and containers
2. If the last rule of the INPUT/FORWARD is a REJECT, concourse generated rules will never be reached.

Not sure if something can be done about the second issue but the first one definitely seems to be a problem.

- concourse 2.7.3 (binary release)
- Linux 4.9.0
- iptables v1.4.21

**Comment by ClaraFuPivotal** [comment]
Could you open this issue on https://github.com/cloudfoundry/guardian? We don't do anything with `iptables` in Concourse itself, so this is out of our control. Thanks!
